rust/src/lib/nm/nm_dbus/connection/route.rs

```diff
@@ -26,20 +33,32 @@ pub struct NmIpRoute {
 impl TryFrom<DbusDictionary> for NmIpRoute {
     type Error = NmError;
     fn try_from(mut v: DbusDictionary) -> Result<Self, Self::Error> {
-        let mut weight = _from_map!(v, "weight", u32::try_from)?;
-        if let Some(weight_num) = weight {
-            if weight_num == 0 {
-                weight = None;
-            }
-        }
+        let dest = _from_map!(v, "dest", String::try_from)?;
+
+        let empty_next_hop_val =
+            match dest.as_deref().unwrap_or("").parse::<std::net::IpAddr>() {
+                Ok(std::net::IpAddr::V4(_)) => IPV4_EMPTY_NEXT_HOP_ADDR,
+                Ok(std::net::IpAddr::V6(_)) => IPV6_EMPTY_NEXT_HOP_ADDR,
```

Comment:
The `nm_dbus` crate should only provide DBUS abstraction without any smart modifications, these smart modification should be done at `nm` level.

```patch
diff --git a/rust/src/lib/nm/query_apply/route.rs b/rust/src/lib/nm/query_apply/route.rs
index d1d6be15..41962016 100644
--- a/rust/src/lib/nm/query_apply/route.rs
+++ b/rust/src/lib/nm/query_apply/route.rs
@@ -3,6 +3,8 @@
 use super::super::nm_dbus::{NmConnection, NmIpRoute};
 
 const DEFAULT_TABLE_ID: u32 = 254; // main route table ID
+const EMPTY_NEXT_HOP_ADDRV4: &str = "0.0.0.0";
+const EMPTY_NEXT_HOP_ADDRV6: &str = "::";
 
 pub(crate) fn is_route_removed(
     new_nm_conn: &NmConnection,
@@ -33,16 +35,39 @@ pub(crate) fn is_route_removed(
     )
 }
 
+fn sanitize_nm_ip_route(route: &mut NmIpRoute) {
+    if route.table.is_none() {
+        route.table = Some(DEFAULT_TABLE_ID);
+    }
+    if route.next_hop.as_deref() == Some(EMPTY_NEXT_HOP_ADDRV4)
+        || route.next_hop.as_deref() == Some(EMPTY_NEXT_HOP_ADDRV6)
+    {
+        route.next_hop = None;
+    }
+}
+
 fn is_nm_ip_route_removed(
     new_routes: &[NmIpRoute],
     cur_routes: &[NmIpRoute],
 ) -> bool {
+    let new_routes: Vec<NmIpRoute> = new_routes
+        .iter()
+        .map(|r| {
+            let mut route = r.clone();
+            sanitize_nm_ip_route(&mut route);
+            route
+        })
+        .collect();
+    let cur_routes: Vec<NmIpRoute> = cur_routes
+        .iter()
+        .map(|r| {
+            let mut route = r.clone();
+            sanitize_nm_ip_route(&mut route);
+            route
+        })
+        .collect();
     for cur_route in cur_routes {
-        let mut cur_route_norm = cur_route.clone();
-        if cur_route_norm.table.is_none() {
-            cur_route_norm.table = Some(DEFAULT_TABLE_ID);
-        }
-        if !new_routes.contains(&cur_route_norm) {
+        if !new_routes.contains(&cur_route) {
             return true;
         }
     }
```